apps/prairielearn/src/lib/bot-grading.ts

```diff
@@ -0,0 +1,182 @@
+  assessment_question: AssessmentQuestion;
+  urlPrefix: string;
+}): Promise<string> {
+  // if OpenAI API Key and Organization are not provided, throw error
+  if (!config.openAiApiKey || !config.openAiOrganization) {
+    throw new error.HttpStatusError(501, 'Not implemented (feature not available)');
+  }
+  const openaiconfig = { apiKey: config.openAiApiKey, organization: config.openAiOrganization };
+
+  const question_course = await getQuestionCourse(question, course);
+
+  const serverJob = await createServerJob({
+    courseId: course.id,
+    courseInstanceId: course_instance_id,
+    assessmentId: assessment_question.assessment_id,
+    authnUserId: '1',
+    type: 'botGrading',
+    description: 'Use LLM to grade assessment question',
+  });
+
+  serverJob.executeInBackground(async (job) => {
+    const openai = new OpenAI(openaiconfig);
+    job.info('OpenAI client ready! ');
+
+    // get all instance questions
+    const result = await queryRows(
+      sql.select_instance_questions_manual_grading,
+      {
+        assessment_id: assessment_question.assessment_id,
+        assessment_question_id: assessment_question.id,
+      },
+      InstanceQuestionSchema,
+    );
+
+    job.info(`Found ${result.length} submissions to grade! `);
+
+    let error_count = 0;
+    let output_count = 0;
+    let output: string | null = null;
+
+    // get each instance question
+    for (const instance_question of result) {
+      const { variant, submission } = await queryRow(
+        sql.select_last_variant_and_submission,
+        { instance_question_id: instance_question.id },
+        SubmissionVariantSchema,
+      );
+
+      // build urls for the question server
+      const urls = buildQuestionUrls(urlPrefix, variant, question, instance_question);
+
+      // get question html
+      const questionModule = questionServers.getModule(question.type);
+      const { courseIssues, data } = await questionModule.render(
+        { question: true, submissions: false, answer: false },
+        variant,
+        question,
+        null,
+        [],
+        question_course,
+        urls,
+      );
+      if (courseIssues.length) {
+        job.info(courseIssues.toString());
+        job.error('Error occurred');
+        job.fail('Errors occurred while bot grading, see output for details');
+      }
+      const question_prompt_raw = data.questionHtml;
+      const $ = cheerio.load(question_prompt_raw, null, false);
+      $('script').remove();
+      const question_prompt = $.html();
+
+      const render_submission_results = await questionModule.render(
+        { question: false, submissions: true, answer: false },
+        variant,
+        question,
+        submission,
+        [submission],
+        question_course,
+        urls,
+      );
+      const student_answer = render_submission_results.data.submissionHtmls[0];
+
+      // Call OpenAI API
+      const completion = await openai.chat.completions.create({
```

Comment:
It's considered a best practice to include end-user IDs: https://platform.openai.com/docs/guides/safety-best-practices/end-user-ids

I'm not sure what the best ID would be here. We could use the user performing the grading? Or treat the entire course as a user?

Reply:
Currently I'm choosing the user performing the grading. Maybe since the user information is there already, we can just use something a little more detailed?

Reply:
Or we can try to include both course id and user id in the API call?

Reply:
Using the course id seems like a reasonable thing to me.

Reply:
Just added the course ID as the user of the API call

Reply:
Let's use `course_${course.id}` (e.g. `course_123`) so that we can differentiate between courses and other entities that may in the future be invoking OpenAI.